sycl/source/detail/kernel_impl.hpp

```diff
@@ -145,6 +145,13 @@ class kernel_impl {
   const ContextImplPtr MContext;
   const ProgramImplPtr MProgramImpl;
   bool MCreatedFromSource = true;
+
+public:
+  // This section means the object is non-movable and non-copyable
+  kernel_impl(const kernel_impl &) = delete;
```

Comment:
I suggest 2 things:
1) to move this section to the place right after the kernel_impl constructors declarations
2) to add a simple comment explaining why they are non-movable. For example:
```suggestion
  // There is no need in move and copy constructors.
  // If they are even needed they must call piKernelRetain method for MKernel.
  kernel_impl(const kernel_impl &) = delete;
```